File: rust/crates/corlinman-gateway/src/legacy_migration.rs
//! Phase 4 W1 4-1A Item 5: one-shot boot migration that moves
//! pre-Phase-4 data files from `<data_dir>/<name>.sqlite` to
//! `<data_dir>/tenants/default/<name>.sqlite`, keeping legacy
//! deployments compatible with the new per-tenant directory layout.
//!
//! The migration is gated on `[tenants].enabled = true &&
//! [tenants].migrate_legacy_paths = true`. When both are off (the
//! default for any pre-Phase-4 config), this module is never called
//! and the gateway continues opening data files at the legacy paths.
//!
//! Idempotency rules:
//!
//! - If `<data_dir>/<name>.sqlite` does not exist, the entry is
//!   skipped (clean install or already-migrated boot).
//! - If `<data_dir>/tenants/default/<name>.sqlite` already exists,
//!   the migration **leaves the legacy file in place** rather than
//!   overwriting. Two files claiming to be the same DB is a manual-
//!   reconcile situation; an automatic rename would silently lose
//!   whichever side has fewer rows. The boot logs both paths at
//!   `tracing::warn` so the operator notices.
//! - The actual move uses `std::fs::rename` which is atomic on the
//!   same filesystem (POSIX `rename(2)`). Cross-device moves return
//!   `std::io::ErrorKind::CrossesDevices` and the migration aborts
//!   that entry with a `tracing::error` — operators must move the
//!   data manually in that case.
//!
//! Files migrated (matches the per-tenant SQLite set Phase 3.1 + Phase
//! 4 Item 1 schema-migrated):
//!
//! - `evolution.sqlite`
//! - `kb.sqlite`
//! - `sessions.sqlite`
//! - `user_model.sqlite`
//! - `agent_state.sqlite`

use std::path::Path;

use corlinman_tenant::tenant_db_path;

/// The five per-tenant SQLite filenames Phase 4 migrates from the
/// legacy flat layout into `<data_dir>/tenants/<tenant>/`.
const LEGACY_DB_NAMES: &[&str] = &[
    "evolution",
    "kb",
    "sessions",
    "user_model",
    "agent_state",
];

/// Walk the legacy data files and rename each into the new
/// `<data_dir>/tenants/default/` layout. Idempotent: see module
/// docs for the rules.
///
/// The function takes `&Path` rather than owning a `PathBuf` so the
/// caller can keep its boot-time `data_dir` value alive across the
/// rest of the runtime construction.
pub fn migrate_legacy_data_files(data_dir: &Path) -> std::io::Result<()> {
    use corlinman_tenant::TenantId;

    let default = TenantId::legacy_default();
    let target_root = data_dir.join("tenants").join(default.as_str());

    // Create the target directory once — `tenant_db_path` does not
    // create parents, and `fs::rename` requires the destination's
    // parent to exist.
    std::fs::create_dir_all(&target_root).map_err(|e| {
        tracing::error!(
            target = %target_root.display(),
            error = %e,
            "failed to create per-tenant data directory; legacy migration aborted",
        );
        e
    })?;

    for name in LEGACY_DB_NAMES {
        let legacy = data_dir.join(format!("{name}.sqlite"));
        let migrated = tenant_db_path(data_dir, &default, name);

        if !legacy.exists() {
            continue;
        }

        if migrated.exists() {
            tracing::warn!(
                legacy = %legacy.display(),
                migrated = %migrated.display(),
                "legacy data file present alongside already-migrated copy; \
                 leaving both untouched — operator should reconcile",
            );
            continue;
        }

        match std::fs::rename(&legacy, &migrated) {
            Ok(()) => {
                tracing::info!(
                    from = %legacy.display(),
                    to = %migrated.display(),
                    "phase 4 legacy data file moved into per-tenant layout",
                );
            }
            Err(e) => {
                tracing::error!(
                    from = %legacy.display(),
                    to = %migrated.display(),
                    error = %e,
                    "phase 4 legacy data file rename failed; \
                     operator must move the file manually",
                );
                // Carry on — a single failed entry should not block
                // the remaining four from migrating.
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::TempDir;

    fn touch(path: &Path, body: &[u8]) {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).unwrap();
        }
        fs::write(path, body).unwrap();
    }

    #[test]
    fn empty_data_dir_is_a_noop() {
        let tmp = TempDir::new().unwrap();
        // Nothing should be created or moved.
        migrate_legacy_data_files(tmp.path()).unwrap();
        // The function does create the target directory tree
        // unconditionally (cheap; idempotent), but no DB files
        // should appear.
        assert!(tmp.path().join("tenants").join("default").is_dir());
        for name in LEGACY_DB_NAMES {
            assert!(!tmp.path().join(format!("{name}.sqlite")).exists());
            assert!(!tmp
                .path()
                .join("tenants")
                .join("default")
                .join(format!("{name}.sqlite"))
                .exists());
        }
    }

    #[test]
    fn moves_every_legacy_file_into_default_tenant_dir() {
        let tmp = TempDir::new().unwrap();
        // Bootstrap a pre-Phase-4 layout: every DB sits at the
        // legacy flat path with a unique payload so we can verify
        // identity-after-rename.
        for name in LEGACY_DB_NAMES {
            touch(
                &tmp.path().join(format!("{name}.sqlite")),
                format!("legacy-{name}").as_bytes(),
            );
        }

        migrate_legacy_data_files(tmp.path()).unwrap();

        for name in LEGACY_DB_NAMES {
            // Source gone.
            assert!(
                !tmp.path().join(format!("{name}.sqlite")).exists(),
                "legacy {name}.sqlite should have been moved",
            );
            // Destination present at the per-tenant path with the
            // same content (rename, not copy).
            let migrated = tmp
                .path()
                .join("tenants")
                .join("default")
                .join(format!("{name}.sqlite"));
            assert!(migrated.exists(), "migrated {name}.sqlite should exist");
            let body = fs::read(&migrated).unwrap();
            assert_eq!(body, format!("legacy-{name}").as_bytes());
        }
    }

    #[test]
    fn already_migrated_path_is_left_in_place() {
        let tmp = TempDir::new().unwrap();
        // Half-migrated state: legacy + new both present. The
        // helper should leave both alone (and warn).
        let legacy = tmp.path().join("evolution.sqlite");
        let migrated = tmp
            .path()
            .join("tenants")
            .join("default")
            .join("evolution.sqlite");
        touch(&legacy, b"legacy-evolution");
        touch(&migrated, b"migrated-evolution");

        migrate_legacy_data_files(tmp.path()).unwrap();

        assert!(legacy.exists(), "legacy file must not be deleted");
        assert!(migrated.exists(), "migrated file must not be deleted");
        assert_eq!(fs::read(&legacy).unwrap(), b"legacy-evolution");
        assert_eq!(fs::read(&migrated).unwrap(), b"migrated-evolution");
    }

    #[test]
    fn idempotent_second_run_after_full_migration() {
        let tmp = TempDir::new().unwrap();
        for name in LEGACY_DB_NAMES {
            touch(
                &tmp.path().join(format!("{name}.sqlite")),
                format!("v1-{name}").as_bytes(),
            );
        }
        migrate_legacy_data_files(tmp.path()).unwrap();
        // Second invocation: the legacy paths are already empty, so
        // the loop short-circuits each entry. No errors, no extra
        // moves.
        migrate_legacy_data_files(tmp.path()).unwrap();
        for name in LEGACY_DB_NAMES {
            assert!(!tmp.path().join(format!("{name}.sqlite")).exists());
            let migrated = tmp
                .path()
                .join("tenants")
                .join("default")
                .join(format!("{name}.sqlite"));
            assert_eq!(fs::read(&migrated).unwrap(), format!("v1-{name}").as_bytes());
        }
    }

    #[test]
    fn partial_legacy_set_only_moves_present_files() {
        let tmp = TempDir::new().unwrap();
        // Only `evolution.sqlite` and `kb.sqlite` exist; the rest
        // are absent. The helper should move just those two.
        touch(&tmp.path().join("evolution.sqlite"), b"e");
        touch(&tmp.path().join("kb.sqlite"), b"k");

        migrate_legacy_data_files(tmp.path()).unwrap();

        let target_root = tmp.path().join("tenants").join("default");
        assert!(target_root.join("evolution.sqlite").exists());
        assert!(target_root.join("kb.sqlite").exists());
        for name in ["sessions", "user_model", "agent_state"] {
            assert!(!target_root.join(format!("{name}.sqlite")).exists());
            assert!(!tmp.path().join(format!("{name}.sqlite")).exists());
        }
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/lifecycle/legacy_migration.py
"""One-shot boot migration: pre-Phase-4 flat data files → per-tenant layout.

Python port of ``rust/crates/corlinman-gateway/src/legacy_migration.rs``.

Moves ``<data_dir>/<name>.sqlite`` into
``<data_dir>/tenants/default/<name>.sqlite`` so deployments running an
older build can upgrade in place without manual file shuffling. Gated by
``[tenants].enabled = true && [tenants].migrate_legacy_paths = true`` on
the Rust side; the Python entrypoint mirrors the same gate before
invoking :func:`migrate_legacy_data_files`.

Idempotency rules (kept byte-for-byte with the Rust impl):

* missing legacy file ⇒ skip silently (clean install / already migrated);
* destination already populated ⇒ leave both untouched and ``warning`` —
  reconciliation is an operator decision, never a silent overwrite;
* rename uses :func:`os.rename`, which is atomic on the same filesystem
  via POSIX ``rename(2)``. Cross-device renames raise
  :class:`OSError` (``EXDEV``); we log and continue with the remaining
  entries rather than abort the boot.

The Rust version uses ``std::fs::rename`` (sync). We keep the Python port
sync as well — the migration runs once at boot, before the event loop is
spun up, and async file I/O would only buy us pretend-concurrency for a
five-element loop. The function is safe to call from inside an async
context (it doesn't block the loop noticeably).
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Final

import structlog

from corlinman_server.tenancy import TenantId, tenant_db_path

logger = structlog.get_logger(__name__)

#: The per-tenant SQLite filenames Phase 4 migrates from the legacy flat
#: layout. Mirrors ``LEGACY_DB_NAMES`` in the Rust crate; keeping the
#: list here so a future per-tenant DB addition only has to touch one
#: place per language.
LEGACY_DB_NAMES: Final[tuple[str, ...]] = (
    "evolution",
    "kb",
    "sessions",
    "user_model",
    "agent_state",
)


def migrate_legacy_data_files(data_dir: Path | str) -> None:
    """Walk the legacy data files and rename each into the new
    per-tenant layout.

    Idempotent: see module docs for the exact rules. Always creates the
    target ``<data_dir>/tenants/default/`` directory (cheap; reused by
    later boot steps that open per-tenant pools), even when no legacy
    files exist — matches the Rust behaviour and keeps subsequent
    ``tenant_db_path`` opens from racing on ``mkdir``.
    """
    root = Path(data_dir)
    default = TenantId.legacy_default()
    target_root = root / "tenants" / default.as_str()

    try:
        target_root.mkdir(parents=True, exist_ok=True)
    except OSError as exc:
        logger.error(
            "legacy_migration.mkdir_failed",
            target=str(target_root),
            error=str(exc),
        )
        raise

    for name in LEGACY_DB_NAMES:
        legacy = root / f"{name}.sqlite"
        migrated = tenant_db_path(root, default, name)

        if not legacy.exists():
            continue

        if migrated.exists():
            logger.warning(
                "legacy_migration.already_migrated",
                legacy=str(legacy),
                migrated=str(migrated),
                detail=(
                    "legacy data file present alongside already-migrated "
                    "copy; leaving both untouched — operator should reconcile"
                ),
            )
            continue

        try:
            os.rename(legacy, migrated)
        except OSError as exc:
            # A single failed entry should not block the remaining four
            # from migrating. EXDEV (cross-device) and EACCES are the
            # usual culprits; both require operator action.
            logger.error(
                "legacy_migration.rename_failed",
                from_path=str(legacy),
                to_path=str(migrated),
                error=str(exc),
                detail=(
                    "phase 4 legacy data file rename failed; operator must "
                    "move the file manually"
                ),
            )
            continue

        logger.info(
            "legacy_migration.moved",
            from_path=str(legacy),
            to_path=str(migrated),
        )


__all__ = ["LEGACY_DB_NAMES", "migrate_legacy_data_files"]
